apis/python/src/tiledbsoma/_read_iters.py

```diff
@@ -5,59 +5,420 @@
 
 """Read iterators.
 """
+from __future__ import annotations
+
 import abc
-from typing import TypeVar
+from concurrent.futures import ThreadPoolExecutor
+from typing import (
+    TYPE_CHECKING,
+    Dict,
+    Iterator,
+    List,
+    Optional,
+    Sequence,
+    Tuple,
+    TypeVar,
+    Union,
+    cast,
+)
 
 import numpy as np
+import numpy.typing as npt
+import pandas as pd
 import pyarrow as pa
 import somacore
+from scipy import sparse
+from somacore import options
+from somacore.query._eager_iter import EagerIterator
 
 # This package's pybind11 code
 import tiledbsoma.pytiledbsoma as clib
 
+from . import _util
+from ._exception import SOMAError
 from ._types import NTuple
 
+if TYPE_CHECKING:
+    from . import SparseNDArray
+
+
+# Convenience types
+_RT = TypeVar("_RT")
+BlockwiseTableReadIterResult = Tuple[pa.Table, Tuple[pa.Array, ...]]
+BlockwiseSingleAxisTableIter = Iterator[BlockwiseTableReadIterResult]
+
+BlockwiseScipyReadIterResult = Tuple[
+    Union[sparse.csr_matrix, sparse.csc_matrix, sparse.coo_matrix],
+    Tuple[npt.NDArray[np.int64], npt.NDArray[np.int64]],
+]
+
+IndicesType = Tuple[npt.NDArray[np.int64], npt.NDArray[np.int64]]
+IJDType = Tuple[
+    Tuple[npt.NDArray[np.int64], npt.NDArray[np.int64]], npt.NDArray[np.generic]
+]
+
 
 class TableReadIter(somacore.ReadIter[pa.Table]):
     """Iterator over `Arrow Table <https://arrow.apache.org/docs/python/generated/pyarrow.Table.html>`_ elements"""
```

Comment:
```suggestion
    """Iterator over `Arrow Table <https://arrow.apache.org/docs/python/generated/pyarrow.Table.html>` elements"""
```

Reply:
This link syntax is  used pervasively, and has something to do with how they are published.  AFAIK, not a typo.

@johnkerl - can you confirm?

Reply:
@bkmartinjr 
https://tiledbsoma.readthedocs.io/en/latest/_autosummary/tiledbsoma.SparseNDArray.html#tiledbsoma.SparseNDArray

Reply:
@johnkerl - not sure I understand the response.  I _believe_ you are saying the underscore is required (something to do with the rendering format is my guess). If you have a doc pointer for the rendering/formatting language, it would be appreciated.

Reply:
The underscore should be there, yes.